aptos-move/aptos-aggregator/src/types.rs

```diff
@@ -37,14 +38,131 @@ pub trait TryFromMoveValue: Sized {
     ) -> Result<Self, Self::Error>;
 }
 
-pub type AggregatorResult<T> = Result<T, AggregatorError>;
+// represents something that should never happen - i.e. a code invariant error,
+// which we would generally just panic, but since we are inside of the VM,
+// we cannot do that.
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub struct PanicError(String);
 
-// TODO: Use this instead of PartialVM errors.
-#[derive(Debug)]
-pub enum AggregatorError {
-    WrongVersionID,
+impl ToString for PanicError {
+    fn to_string(&self) -> String {
+        self.0.clone()
+    }
+}
+
+// Wrapping another error, to add a variant that represents
+// something that should never happen - i.e. a code invariant error,
+// which we would generally just panic, but since we are inside of the VM,
+// we cannot do that.
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub enum PanicOr<T: std::fmt::Debug> {
+    CodeInvariantError(String),
+    Or(T),
+}
+
+pub type PanicOrResult<T, E> = Result<T, PanicOr<E>>;
+
+pub fn code_invariant_error<M: std::fmt::Debug>(message: M) -> PanicError {
+    let msg = format!(
+        "Delayed logic code invariant broken (there is a bug in the code), {:?}",
+        message
+    );
+    error!("{}", msg);
+    PanicError(msg)
+}
+
+pub fn expect_ok<V, E: std::fmt::Debug>(value: Result<V, E>) -> Result<V, PanicError> {
+    value.map_err(code_invariant_error)
+}
+
+impl<T: std::fmt::Debug> From<PanicError> for PanicOr<T> {
+    fn from(err: PanicError) -> Self {
+        PanicOr::CodeInvariantError(err.0)
+    }
+}
+
+pub trait NonPanic {}
+// impl NonPanic for f64 {}
+
+impl<T: std::fmt::Debug + NonPanic> From<T> for PanicOr<T> {
+    fn from(err: T) -> Self {
+        PanicOr::Or(err)
+    }
 }
 
+impl From<PanicError> for PartialVMError {
+    fn from(err: PanicError) -> Self {
+        PartialVMError::from(PanicOr::<()>::from(err))
+    }
+}
+
+impl From<DelayedFieldsSpeculativeError> for PartialVMError {
+    fn from(err: DelayedFieldsSpeculativeError) -> Self {
+        PartialVMError::from(PanicOr::from(err))
+    }
+}
+
+impl<T: std::fmt::Debug> From<&PanicOr<T>> for StatusCode {
+    fn from(err: &PanicOr<T>) -> Self {
+        match err {
+            PanicOr::CodeInvariantError(_) => StatusCode::DELAYED_FIELDS_CODE_INVARIANT_ERROR,
```

Comment:
Is this right? Don't we want invariant code violation code here?

Reply:
INVARIANT_CODE_VIOLATION is something that can get committed on chain

our error we want to check inside of BlockSTM, and fallback to sequential execution - if it occurs.

Reply:
Got it, can you add a comment saying this haha